Storage/SQL/SQLiteHistory.swift

```diff
@@ -105,6 +148,214 @@ open class SQLiteHistory {
+ */
+fileprivate struct FrecentHistoryImpl : FrecentHistory {
+    private let db: BrowserDB
+    private let prefs: Prefs
+
+    init(db: BrowserDB, prefs: Prefs) {
+        self.db = db
+        self.prefs = prefs
+
+        let create =
+            "CREATE TEMP TABLE \(AwesomebarBookmarksTempTable) (" +
+                "guid TEXT, url TEXT, title TEXT, description TEXT, " +
+        "faviconID INT, visitDate DATE)"
+        let insert =
+            "INSERT INTO \(AwesomebarBookmarksTempTable) " +
+                "SELECT b.guid AS guid, b.url AS url, b.title AS title, " +
+                "b.description AS description, b.faviconID AS faviconID, " +
+                "h.visitDate AS visitDate " +
+                "FROM \(ViewAwesomebarBookmarks) b " +
+        "LEFT JOIN \(ViewHistoryVisits) h ON b.url = h.url"
+
+        db.withConnection { connection in
+            let existsQuery = "SELECT COUNT(*) FROM sqlite_temp_master WHERE type = 'table' and name = '\(AwesomebarBookmarksTempTable)'"
+            let tableExists = 1 == connection.executeQuery(existsQuery, factory: IntFactory).asArray().first
+            if tableExists {
+                try connection.executeChange("DELETE FROM \(AwesomebarBookmarksTempTable)")
+            } else {
+                try connection.executeChange(create)
+            }
+            try connection.executeChange(insert)
+        }
+    }
+
+    func getSites(whereURLContains filter: String?, historyLimit limit: Int, bookmarksLimit: Int) -> Deferred<Maybe<Cursor<Site>>> {
+        let factory = SQLiteHistory.basicHistoryColumnFactory
+
+        let params = FrecencyParams.urlCompletion(bookmarksLimit: bookmarksLimit, whereURLContains: filter ?? "", groupClause: "GROUP BY historyID ")
+        let (query, args) = getFrecencyQuery(historyLimit: limit, params: params)
+
+        return db.runQuery(query, args: args, factory: factory)
+    }
+
+    func updateTopSitesCacheQuery() -> (String, Args?) {
+        let limit = Int(prefs.intForKey(PrefsKeys.KeyTopSitesCacheSize) ?? TopSiteCacheSize)
+
+        let (whereData, groupBy) = topSiteClauses()
+        let params = FrecencyParams.topSites(groupClause: groupBy, whereData: whereData)
+        let (frecencyQuery, args) = getFrecencyQuery(historyLimit: limit, params: params)
+
+        // We must project, because we get bookmarks in these results.
+        let insertQuery = [
+            "WITH siteFrecency AS (\(frecencyQuery))",
+            "INSERT INTO \(TableCachedTopSites)",
+            "SELECT historyID, url, title, guid, domain_id, domain,",
+            "localVisitDate, remoteVisitDate, localVisitCount, remoteVisitCount,",
+            "iconID, iconURL, iconDate, iconWidth, frecencies",
+            "FROM (SELECT * FROM",
+            "siteFrecency LEFT JOIN view_history_id_favicon ON",
+            "siteFrecency.historyID = view_history_id_favicon.id)"
+            ].joined(separator: " ")
```

Comment:
I expect that `swiftc` has fixed the unpleasant concatenation compile time bug that prompted us to do this. Oh, for multi-line strings.